File: ios/Classes/AnylineTireTreadPlugin.swift
import Flutter
import UIKit
import AnylineTireTreadSdk

public class AnylineTireTreadPlugin: NSObject, FlutterPlugin {

    private var pendingScanResult: FlutterResult?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "anyline_tire_tread_plugin",
                                           binaryMessenger: registrar.messenger())
        let instance = AnylineTireTreadPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case Constants.METHOD_GET_SDK_VERSION:
            result(AnylineTireTread.shared.sdkVersion)
        case Constants.METHOD_INITIALIZE:
            initializeSdk(result: result, call: call)
        case Constants.METHOD_SCAN:
            scan(result: result, call: call)
        case Constants.METHOD_IS_DEVICE_SUPPORTED:
            isDeviceSupported(result: result)
        case Constants.METHOD_GET_RESULT:
            getResult(result: result, call: call)
        case Constants.METHOD_GET_HEATMAP:
            getHeatmap(result: result, call: call)
        case Constants.METHOD_SEND_FEEDBACK_COMMENT:
            sendCommentFeedback(result: result, call: call)
        case Constants.METHOD_SEND_TREAD_DEPTH_RESULT_FEEDBACK:
            sendTreadDepthResultFeedback(result: result, call: call)
        case Constants.METHOD_SEND_TIRE_ID_FEEDBACK:
            sendTireIdFeedback(result: result, call: call)
        case Constants.METHOD_SET_EXPERIMENTAL_FLAGS:
            setExperimentalFlags(result: result, call: call)
        case Constants.METHOD_CLEAR_EXPERIMENTAL_FLAGS:
            clearExperimentalFlags(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - SDK Methods

    private func initializeSdk(result: @escaping FlutterResult, call: FlutterMethodCall) {
        let arguments = call.arguments as? [String: Any]
        guard let licenseKey = arguments?[Constants.EXTRA_LICENSE_KEY] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "License key parameter not found", details: nil))
            return
        }
        let pluginVersion = arguments?[Constants.EXTRA_PLUGIN_VERSION] as? String ?? ""
        let rawCustomTag = (arguments?[Constants.EXTRA_CUSTOM_TAG] as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let customTag = (rawCustomTag?.isEmpty ?? true) ? nil : rawCustomTag

        AnylineTireTread.shared.initialize(
            licenseKey: licenseKey,
            options: InitOptions(
                customTag: customTag,
                wrapperInfo: WrapperInfo.Flutter(version: pluginVersion)
            )
        ) { sdkResult in
            DispatchQueue.main.async {
                if sdkResult.isOk {
                    result(true)
                } else if let error = sdkResult.error {
                    result(FlutterError(
                        code: error.code.name,
                        message: error.message,
                        details: error.debug?.description
                    ))
                }
            }
        }
    }

    private func scan(result: @escaping FlutterResult, call: FlutterMethodCall) {
        if pendingScanResult != nil {
            result(failedOutcome(code: .alreadyRunning, message: "A scan is already in progress"))
            return
        }

        // Present from the topmost view controller (parity with the RN
        // wrapper's RCTPresentedViewController): presenting from the root VC
        // fails when another controller is already presented.
        guard let presenter = topViewController() else {
            result(failedOutcome(
                code: .invalidArgument,
                message: "Unable to present scan: no active view controller"))
            return
        }

        pendingScanResult = result

        let arguments = call.arguments as? [String: Any]
        let configJson = arguments?[Constants.EXTRA_CONFIG_JSON] as? String
        let optionsJson = arguments?[Constants.EXTRA_SCAN_OPTIONS_JSON] as? String

        AnylineTireTreadScanner().scan(
            from: presenter,
            configJson: configJson,
            optionsJson: optionsJson
        ) { [weak self] outcome in
            DispatchQueue.main.async {
                guard let pending = self?.pendingScanResult else { return }
                self?.pendingScanResult = nil
                pending(self?.sanitizedOutcome(outcome) ?? [:])
            }
        }
    }

    /// Returns the topmost presented view controller of the key window.
    private func topViewController() -> UIViewController? {
        let windows = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
        let window = windows.first { $0.isKeyWindow } ?? windows.first
        var top = window?.rootViewController
        while let presented = top?.presentedViewController {
            top = presented
        }
        return top
    }

    private func failedOutcome(code: ErrorCode, message: String) -> [String: Any] {
        let error = SdkError(
            code: code,
            type: ErrorType.companion.fromCode(code: code),
            message: message,
            debug: nil
        )
        return sanitizedOutcome(FailedOutcome(measurementUUID: nil, error: error))
    }

    private func isDeviceSupported(result: @escaping FlutterResult) {
        AnylineTireTread.shared.isDeviceSupported { sdkResult in
            DispatchQueue.main.async {
                if let supported = sdkResult.result as? Bool {
                    result(supported)
                } else if let error = sdkResult.error {
                    result(FlutterError(
                        code: error.code.name,
                        message: error.message,
                        details: error.debug?.description
                    ))
                }
            }
        }
    }

    private func getResult(result: @escaping FlutterResult, call: FlutterMethodCall) {
        let arguments = call.arguments as? [String: Any]
        guard let uuid = arguments?[Constants.EXTRA_MEASUREMENT_UUID] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "uuid parameter not found", details: nil))
            return
        }
        let timeout = (arguments?[Constants.EXTRA_TIMEOUT_SECONDS] as? NSNumber)?.int32Value ?? 60

        AnylineTireTread.shared.getResult(
            measurementUUID: uuid,
            timeoutSeconds: timeout
        ) { sdkResult in
            DispatchQueue.main.async {
                if let treadResult = sdkResult.result {
                    result(treadResult.toJson())
                } else if let error = sdkResult.error {
                    result(FlutterError(
                        code: error.code.name,
                        message: error.message,
                        details: error.debug?.description
                    ))
                }
            }
        }
    }

    private func getHeatmap(result: @escaping FlutterResult, call: FlutterMethodCall) {
        let arguments = call.arguments as? [String: Any]
        guard let uuid = arguments?[Constants.EXTRA_MEASUREMENT_UUID] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "uuid parameter not found", details: nil))
            return
        }
        let timeout = (arguments?[Constants.EXTRA_TIMEOUT_SECONDS] as? NSNumber)?.int32Value ?? 60

        AnylineTireTread.shared.getHeatmap(
            measurementUUID: uuid,
            timeoutSeconds: timeout
        ) { sdkResult in
            DispatchQueue.main.async {
                if let heatmap = sdkResult.result {
                    result(heatmap.url)
                } else if let error = sdkResult.error {
                    result(FlutterError(
                        code: error.code.name,
                        message: error.message,
                        details: error.debug?.description
                    ))
                }
            }
        }
    }

    private func sendCommentFeedback(result: @escaping FlutterResult, call: FlutterMethodCall) {
        let arguments = call.arguments as? [String: Any]
        guard let uuid = arguments?[Constants.EXTRA_MEASUREMENT_UUID] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "uuid parameter not found", details: nil))
            return
        }
        guard let comment = arguments?[Constants.EXTRA_FEEDBACK_COMMENT] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "comment parameter not found", details: nil))
            return
        }

        AnylineTireTread.shared.sendCommentFeedback(
            measurementUUID: uuid,
            comment: comment
        ) { sdkResult in
            DispatchQueue.main.async {
                if let info = sdkResult.result {
                    result(self.measurementInfoJson(info))
                } else if let error = sdkResult.error {
                    result(FlutterError(
                        code: error.code.name,
                        message: error.message,
                        details: nil
                    ))
                }
            }
        }
    }

    private func sendTreadDepthResultFeedback(result: @escaping FlutterResult, call: FlutterMethodCall) {
        let arguments = call.arguments as? [String: Any]
        guard let uuid = arguments?[Constants.EXTRA_MEASUREMENT_UUID] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "uuid parameter not found", details: nil))
            return
        }
        guard let regionsString = arguments?[Constants.EXTRA_TREAD_DEPTH_RESULT_FEEDBACK] as? String,
              let data = regionsString.data(using: .utf8),
              let jsonArray = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "regions parameter not found", details: nil))
            return
        }

        let regions: [TreadResultRegion] = jsonArray.map { region in
            let available = region["available"] as? Bool ?? false
            let valueMm = region["value_mm"] as? Double ?? 0.0
            return TreadResultRegion.companion.doInitMm(isAvailable: available, value: valueMm)
        }

        AnylineTireTread.shared.sendTreadDepthResultFeedback(
            measurementUUID: uuid,
            treadResultRegions: regions
        ) { sdkResult in
            DispatchQueue.main.async {
                if let info = sdkResult.result {
                    result(self.measurementInfoJson(info))
                } else if let error = sdkResult.error {
                    result(FlutterError(
                        code: error.code.name,
                        message: error.message,
                        details: nil
                    ))
                }
            }
        }
    }

    private func sendTireIdFeedback(result: @escaping FlutterResult, call: FlutterMethodCall) {
        let arguments = call.arguments as? [String: Any]
        guard let uuid = arguments?[Constants.EXTRA_MEASUREMENT_UUID] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "uuid parameter not found", details: nil))
            return
        }
        guard let tireId = arguments?[Constants.EXTRA_TIRE_ID] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "tireId parameter not found", details: nil))
            return
        }

        AnylineTireTread.shared.sendTireIdFeedback(
            measurementUUID: uuid,
            tireId: tireId
        ) { sdkResult in
            DispatchQueue.main.async {
                if let info = sdkResult.result {
                    result(self.measurementInfoJson(info))
                } else if let error = sdkResult.error {
                    result(FlutterError(
                        code: error.code.name,
                        message: error.message,
                        details: nil
                    ))
                }
            }
        }
    }

    private func setExperimentalFlags(result: @escaping FlutterResult, call: FlutterMethodCall) {
        let arguments = call.arguments as? [String: Any]
        guard let flags = arguments?[Constants.EXTRA_EXPERIMENTAL_FLAGS] as? [String] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Experimental flags are null", details: nil))
            return
        }
        AnylineTireTreadSdk.shared.setExperimentalFlags(newFlags: flags)
        result(true)
    }

    private func clearExperimentalFlags(result: @escaping FlutterResult) {
        AnylineTireTreadSdk.shared.clearExperimentalFlags()
        result(true)
    }

    // MARK: - Helpers

    /// Serializes a [MeasurementInfo] to the JSON shape the Dart
    /// `MeasurementInfo.fromJson` expects. The SDK type exposes no toJson(),
    /// so we build it from its typed properties.
    private func measurementInfoJson(_ info: MeasurementInfo) -> String? {
        var map: [String: Any] = [
            "measurementUUID": info.measurementUUID,
            "status": info.status.name,
        ]
        if let context = info.additionalContext {
            var contextMap: [String: Any] = [:]
            if let correlationId = context.correlationId {
                contextMap["correlationId"] = correlationId
            }
            if let position = context.tirePosition {
                contextMap["tirePosition"] = [
                    "axle": Int(position.axle),
                    "positionOnAxle": Int(position.positionOnAxle),
                    "side": position.side.name,
                ]
            }
            if !contextMap.isEmpty {
                map["additionalContext"] = contextMap
            }
        }
        guard let data = try? JSONSerialization.data(withJSONObject: map),
              let json = String(data: data, encoding: .utf8) else {
            return nil
        }
        return json
    }

    private func sanitizedOutcome(_ outcome: ScanOutcome) -> [String: Any] {
        // Use the SDK's Bridge to obtain a Flutter-codec-safe map, matching the
        // Android implementation. The model's own toMap() embeds non-codec
        // types on the failure path — notably SdkError.debug, a map of kotlinx
        // JsonElement objects — which crash StandardMessageCodec when the reply
        // is encoded. Bridge flattens everything to Foundation primitives.
        let raw = Bridge.shared.outcome(scanOutcome: outcome)
        return raw.filter { Self.allowedOutcomeKeys.contains($0.key) }
    }

    private static let allowedOutcomeKeys: Set<String> = ["kind", "measurementUUID", "error"]
}

File: android/src/main/kotlin/io/anyline/tiretread/flutter/AnylineTireTreadPlugin.kt
package io.anyline.tiretread.flutter

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.anyline.tiretread.sdk.AnylineTireTreadSdk
import io.anyline.tiretread.sdk.InternalAPI
import io.anyline.tiretread.sdk.api.AnylineTireTread
import io.anyline.tiretread.sdk.api.AnylineTireTreadScanner
import io.anyline.tiretread.sdk.api.Bridge
import io.anyline.tiretread.sdk.api.ErrorCode
import io.anyline.tiretread.sdk.api.FailedOutcome
import io.anyline.tiretread.sdk.api.InitOptions
import io.anyline.tiretread.sdk.api.ScanOutcome
import io.anyline.tiretread.sdk.api.SdkError
import io.anyline.tiretread.sdk.api.SdkResult
import io.anyline.tiretread.sdk.types.MeasurementInfo
import io.anyline.tiretread.sdk.types.TreadResultRegion
import io.anyline.tiretread.sdk.types.WrapperInfo
import org.json.JSONObject

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

class AnylineTireTreadPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var applicationContext: Context? = null
    private var activityPluginBinding: ActivityPluginBinding? = null
    private var pendingScanResult: Result? = null

    // The Tire Sidewall (TSW) scanner is exposed over its own channel. A Flutter
    // package declares a single pluginClass per platform, so this plugin owns
    // the sidewall handler and forwards the FlutterPlugin/ActivityAware lifecycle.
    private val sidewallPlugin = AnylineTireSidewallPlugin()

    // SDK v15 invokes its callbacks on a background dispatcher (Dispatchers.IO).
    // MethodChannel results must be delivered on the platform (main) thread, so
    // every SDK callback is marshalled through this handler. Lazy so the class
    // can still be instantiated in JVM unit tests without a prepared Looper.
    private val mainHandler by lazy { Handler(Looper.getMainLooper()) }

    private fun onMain(block: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            block()
        } else {
            mainHandler.post(block)
        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "anyline_tire_tread_plugin")
        channel.setMethodCallHandler(this)
        sidewallPlugin.onAttachedToEngine(flutterPluginBinding)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            Constants.METHOD_INITIALIZE -> {
                val licenseKey = call.argument<String>(Constants.EXTRA_LICENSE_KEY) ?: ""
                val pluginVersion = call.argument<String>(Constants.EXTRA_PLUGIN_VERSION) ?: ""
                val customTag = call.argument<String>(Constants.EXTRA_CUSTOM_TAG)
                initializeSdk(licenseKey, pluginVersion, customTag, result)
            }
            Constants.METHOD_GET_SDK_VERSION -> {
                result.success(AnylineTireTread.sdkVersion)
            }
            Constants.METHOD_SCAN -> {
                val configJson = call.argument<String>(Constants.EXTRA_CONFIG_JSON)
                val optionsJson = call.argument<String>(Constants.EXTRA_SCAN_OPTIONS_JSON)
                scan(configJson, optionsJson, result)
            }
            Constants.METHOD_IS_DEVICE_SUPPORTED -> {
                isDeviceSupported(result)
            }
            Constants.METHOD_GET_RESULT -> {
                val uuid = call.argument<String>(Constants.EXTRA_MEASUREMENT_UUID) ?: ""
                val timeout = call.argument<Int>(Constants.EXTRA_TIMEOUT_SECONDS)
                getResult(uuid, timeout, result)
            }
            Constants.METHOD_GET_HEATMAP -> {
                val uuid = call.argument<String>(Constants.EXTRA_MEASUREMENT_UUID) ?: ""
                val timeout = call.argument<Int>(Constants.EXTRA_TIMEOUT_SECONDS)
                getHeatmap(uuid, timeout, result)
            }
            Constants.METHOD_SEND_FEEDBACK_COMMENT -> {
                val uuid = call.argument<String>(Constants.EXTRA_MEASUREMENT_UUID) ?: ""
                val comment = call.argument<String>(Constants.EXTRA_FEEDBACK_COMMENT) ?: ""
                sendCommentFeedback(uuid, comment, result)
            }
            Constants.METHOD_SEND_TREAD_DEPTH_RESULT_FEEDBACK -> {
                val uuid = call.argument<String>(Constants.EXTRA_MEASUREMENT_UUID) ?: ""
                val regionsJson = call.argument<String>(Constants.EXTRA_TREAD_DEPTH_RESULT_FEEDBACK) ?: "[]"
                sendTreadDepthResultFeedback(uuid, parseFeedbackRegions(regionsJson), result)
            }
            Constants.METHOD_SEND_TIRE_ID_FEEDBACK -> {
                val uuid = call.argument<String>(Constants.EXTRA_MEASUREMENT_UUID) ?: ""
                val tireId = call.argument<String>(Constants.EXTRA_TIRE_ID) ?: ""
                sendTireIdFeedback(uuid, tireId, result)
            }
            Constants.METHOD_SET_EXPERIMENTAL_FLAGS -> {
                setExperimentalFlags(call, result)
            }
            Constants.METHOD_CLEAR_EXPERIMENTAL_FLAGS -> {
                clearExperimentalFlags(result)
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        applicationContext = null
        sidewallPlugin.onDetachedFromEngine(binding)
    }

    // ActivityAware

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityPluginBinding = binding
        sidewallPlugin.onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityPluginBinding = null
        sidewallPlugin.onDetachedFromActivityForConfigChanges()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activityPluginBinding = binding
        sidewallPlugin.onReattachedToActivityForConfigChanges(binding)
    }

    override fun onDetachedFromActivity() {
        activityPluginBinding = null
        sidewallPlugin.onDetachedFromActivity()
    }

    /// Prefer the current Activity; fall back to the application context
    /// (matches the React Native wrapper's `currentActivity ?: context`).
    private fun hostContext(): Context? =
        activityPluginBinding?.activity ?: applicationContext

    // SDK methods

    private fun initializeSdk(
        licenseKey: String, pluginVersion: String, customTag: String?, result: Result
    ) {
        val context = hostContext()
        if (context == null) {
            result.error(ErrorCode.INVALID_ARGUMENT.name, "Plugin is not attached to a context", null)
            return
        }

        AnylineTireTread.initialize(
            context = context,
            licenseKey = licenseKey,
            options = InitOptions(
                customTag = customTag?.trim().takeUnless { it.isNullOrEmpty() },
                wrapperInfo = WrapperInfo.Flutter(pluginVersion)
            )
        ) { sdkResult ->
            onMain {
                when (sdkResult) {
                    is SdkResult.Ok -> result.success(true)
                    is SdkResult.Err -> result.error(
                        sdkResult.error.code.name,
                        sdkResult.error.message,
                        sdkResult.error.debug?.toString()
                    )
                }
            }
        }
    }

    private fun scan(configJson: String?, optionsJson: String?, result: Result) {
        val context = hostContext()
        if (context == null) {
            val outcome = FailedOutcome(
                measurementUUID = null,
                error = SdkError(
                    code = ErrorCode.INVALID_ARGUMENT,
                    message = "Plugin is not attached to a context"
                )
            )
            result.success(sanitizedOutcome(outcome))
            return
        }

        if (pendingScanResult != null) {
            val outcome = FailedOutcome(
                measurementUUID = null,
                error = SdkError(
                    code = ErrorCode.ALREADY_RUNNING,
                    message = "A scan is already in progress"
                )
            )
            result.success(sanitizedOutcome(outcome))
            return
        }

        pendingScanResult = result

        AnylineTireTreadScanner().scan(
            from = context,
            configJson = configJson,
            optionsJson = optionsJson,
        ) { outcome ->
            onMain {
                val pending = pendingScanResult ?: return@onMain
                pendingScanResult = null
                pending.success(sanitizedOutcome(outcome))
            }
        }
    }

    private fun isDeviceSupported(result: Result) {
        val context = hostContext()
        if (context == null) {
            result.error(
                ErrorCode.INVALID_ARGUMENT.name,
                "Plugin is not attached to a context",
                null
            )
            return
        }
        AnylineTireTread.isDeviceSupported(context) { sdkResult ->
            onMain {
                when (sdkResult) {
                    is SdkResult.Ok -> result.success(sdkResult.result)
                    is SdkResult.Err -> result.error(
                        sdkResult.error.code.name,
                        sdkResult.error.message,
                        sdkResult.error.debug?.toString()
                    )
                }
            }
        }
    }

    private fun getResult(uuid: String, timeoutSeconds: Int?, result: Result) {
        AnylineTireTread.getResult(
            measurementUUID = uuid,
            timeoutSeconds = timeoutSeconds ?: 60,
        ) { sdkResult ->
            onMain {
                when (sdkResult) {
                    is SdkResult.Ok -> result.success(Json.encodeToString(sdkResult.result))
                    is SdkResult.Err -> result.error(
                        sdkResult.error.code.name,
                        sdkResult.error.message,
                        sdkResult.error.debug?.toString()
                    )
                }
            }
        }
    }

    private fun getHeatmap(uuid: String, timeoutSeconds: Int?, result: Result) {
        AnylineTireTread.getHeatmap(
            measurementUUID = uuid,
            timeoutSeconds = timeoutSeconds ?: 60,
        ) { sdkResult ->
            onMain {
                when (sdkResult) {
                    is SdkResult.Ok -> result.success(sdkResult.result.url)
                    is SdkResult.Err -> result.error(
                        sdkResult.error.code.name,
                        sdkResult.error.message,
                        sdkResult.error.debug?.toString()
                    )
                }
            }
        }
    }

    private fun sendCommentFeedback(uuid: String, comment: String, result: Result) {
        AnylineTireTread.sendCommentFeedback(
            measurementUUID = uuid,
            comment = comment
        ) { sdkResult ->
            onMain {
                when (sdkResult) {
                    is SdkResult.Ok -> result.success(measurementInfoJson(sdkResult.result))
                    is SdkResult.Err -> result.error(
                        sdkResult.error.code.name,
                        sdkResult.error.message,
                        null
                    )
                }
            }
        }
    }

    private fun sendTreadDepthResultFeedback(
        uuid: String, regions: List<TreadResultRegion>, result: Result
    ) {
        AnylineTireTread.sendTreadDepthResultFeedback(
            measurementUUID = uuid,
            treadResultRegions = regions
        ) { sdkResult ->
            onMain {
                when (sdkResult) {
                    is SdkResult.Ok -> result.success(measurementInfoJson(sdkResult.result))
                    is SdkResult.Err -> result.error(
                        sdkResult.error.code.name,
                        sdkResult.error.message,
                        null
                    )
                }
            }
        }
    }

    private fun sendTireIdFeedback(uuid: String, tireId: String, result: Result) {
        AnylineTireTread.sendTireIdFeedback(
            measurementUUID = uuid,
            tireId = tireId
        ) { sdkResult ->
            onMain {
                when (sdkResult) {
                    is SdkResult.Ok -> result.success(measurementInfoJson(sdkResult.result))
                    is SdkResult.Err -> result.error(
                        sdkResult.error.code.name,
                        sdkResult.error.message,
                        null
                    )
                }
            }
        }
    }

    @OptIn(InternalAPI::class)
    private fun setExperimentalFlags(call: MethodCall, result: Result) {
        val flags: List<String>? = call.argument(Constants.EXTRA_EXPERIMENTAL_FLAGS)
        if (flags != null) {
            AnylineTireTreadSdk.setExperimentalFlags(newFlags = flags)
            result.success(true)
        } else {
            result.error(ErrorCode.INVALID_ARGUMENT.name, "Experimental flags are null", null)
        }
    }

    @OptIn(InternalAPI::class)
    private fun clearExperimentalFlags(result: Result) {
        AnylineTireTreadSdk.clearExperimentalFlags()
        result.success(true)
    }

    /// Builds feedback regions from `available` + `value_mm` only, via
    /// `TreadResultRegion.initMm` — mirroring the RN wrapper and the iOS
    /// implementation. Deserializing the full region JSON would let the
    /// `value_inch_32nds` setter re-quantize `value_mm` and lose precision.
    private fun parseFeedbackRegions(regionsJson: String): List<TreadResultRegion> =
        Json.parseToJsonElement(regionsJson).jsonArray.map { element ->
            val region = element.jsonObject
            TreadResultRegion.initMm(
                isAvailable = region["available"]?.jsonPrimitive?.booleanOrNull ?: false,
                value = region["value_mm"]?.jsonPrimitive?.doubleOrNull ?: 0.0
            )
        }

    private fun sanitizedOutcome(outcome: ScanOutcome): Map<String, Any?> {
        val raw = Bridge.outcome(outcome)
        return raw.filterKeys { it in OUTCOME_KEYS }
    }

    /// Serializes a [MeasurementInfo] to the JSON shape the Dart
    /// `MeasurementInfo.fromJson` expects, built from its typed properties.
    private fun measurementInfoJson(info: MeasurementInfo): String {
        val json = JSONObject()
        json.put("measurementUUID", info.measurementUUID)
        json.put("status", info.status.name)
        info.additionalContext?.let { context ->
            val contextJson = JSONObject()
            context.correlationId?.let { contextJson.put("correlationId", it) }
            context.tirePosition?.let { position ->
                contextJson.put(
                    "tirePosition",
                    JSONObject()
                        .put("axle", position.axle)
                        .put("positionOnAxle", position.positionOnAxle)
                        .put("side", position.side.name)
                )
            }
            if (contextJson.length() > 0) {
                json.put("additionalContext", contextJson)
            }
        }
        return json.toString()
    }

    companion object {
        private val OUTCOME_KEYS = setOf("kind", "measurementUUID", "error")
    }
}
